rfcs/AP-RFC-003.md

```diff
@@ -0,0 +1,68 @@
+
+- **Robust Artifact Management**: Agents will have a more resilient and flexible system to handle artifacts, thereby reducing potential errors and inefficiencies.
+- **Clear Artifact Source Identification**: It will be easier to trace the source of artifacts, enhancing traceability and auditing capabilities.
+- **Unopinionated**: Unopinionated agent class design will allow developers to implement their desired artifact handling mechanisms.
+
+## Design Proposal
+
+1. **Flexible Upload Paths**: 
+    - Allow agents to specify where files are uploaded from, accommodating external storage solutions such as AWS S3.
+    - Make relative path an optional parameter and mark it to be removed
+    - Add a URI parameter as an optional parameter
+    - Make file an optional parameter
+    - Require either URI or file to be specified
+    - Add an endpoint to the agent protocol to list the URI protocols that are valid for input and output for agents
+
+2. **Artifact Source Identification**:
+    - Introduce a mechanism, possibly via metadata, to discern if an artifact is generated by the agent or if it's an input.
+    
+3. **Revise Current Implementation**:
+    - Modify POST /agent/tasks/<task_id>/artifacts’s file parameter type from string to multipart/form-data 
+    - Remove current artifact handling logic from agent class
+    - Require a handler function in the same way the task and step handler functions are required. This will allow agent developers to implement their desired artifact handling mechanisms.
+
+
+### Alternatives Considered
+
+- **Maintain Current Implementation**: While this ensures stability, it doesn't address the limitations faced by the agent developers.
+- **Batteries Included Approach**: Add the to the protocol a complete artifact handling mechanism. This will make the protocol more opinionated and more complex. Moving away from the original design goals of providing an protocol rather than an implementation.
+
+### Compatibility
+
+- Changes to the artifact class will only add additional fields with defaults set to preserve backward compatibility.
+- Removing the current artifact handling logic will break compatibility with existing agents.
+
+## Detailed Design (Optional)
+
+1. **Enhanced Artifact Class**:
+    - Retain the existing `Artifact` and `ArtifactUpload` classes.
+    - Incorporate attributes to identify artifact source (e.g., `sourceType` with possible values: 'input' or 'output').
+    - Provide methods/functions to specify external upload paths.
+  
+2. **Remove Artifact and Workspace Handling form Agent Class**:
+    - Remove the opinionated artifact and workspace handling logic from the agent class.
+
+3. **Require artifact upload and download handler functions**: 
+    - Require handler functions for artifact upload and download, which can be called form the protocol artifact endpoints.
+
+## Questions and Discussion Topics
+
+- Do we want to keep the artifact handling logic out of the agent protocol?
```

Comment:
I am more inclined to keep it out, only the interface should be given. Let's keep it unopinionated.